include/envoy/upstream/upstream.h

```diff
@@ -646,6 +646,25 @@ using ProtocolOptionsConfigConstSharedPtr = std::shared_ptr<const ProtocolOption
  */
 class ClusterTypedMetadataFactory : public Envoy::Config::TypedMetadataFactory {};
 
+class ConnectionRequestPolicySubscriber {
```

Comment:
Can you add interface comments for these classes and functions? It's not immediately clear to me why these policies need to live in upstream.h. Can we potentially move them to a new file? I'm not sure how they are going to get configured.

Reply:
I have added comments and TODOs hoping to make the intent clearer. There is no strong reason for having the policy interfaces in `upstream.h`. The only reason i went with it for now is due to its proximity to `ClusterInfo`. Since the policies are related (1:1) to ClusterInfo,  they could be declared together. I am open to suggestions.

Reply:
OK thanks this is more clear. IMO I would simplify this for right now and not worry about a public interface, cluster info, etc. I would probably just build this config directly into the a new http2 connection pool options config: https://github.com/envoyproxy/envoy/blob/e03936eaa0c1069a29ab42cb5d8f5cec675c1ee9/api/envoy/api/v2/cds.proto#L252. You can then plumb that config directly from cluster info -> the http2 connection pool much like we do for the existing http2 protocol options.

In terms of these interfaces and implementation, I would either just define them right now directly in the http2 conn pool code, or make new files in that area to contain them. We can always make this more generic later but IMO it's fine to define this where we are going to use this for now.

Reply:
@mattklein123 Thanks! How do you visualize the `Htt2ConnectionPolicy` config? I suppose something like  
```
 clusters:                                                                      
    - name: cluster-name                                                   
      connect_timeout: 0.25s       
      connection_policy:
           max_requests_per_connection: 1  
           drain_on_overflow: yes // Should drain or keep in overflow list
           idle_timeout: 500ms     // Time after which the connection can be closed.                   
                                                  
      http2_protocol_options: {...}
      ...       
                   
```

Reply:
Something like this yes, but per @alyssawilk we should see if we can also unify with HTTP/1.